### PR TITLE
Email user on pipeline completion

### LINKED ISSUE #80: Email on pipeline completion
## Describe your idea for a new feature
Users would like the option to get an email notification when their pipeline is complete.  

## Describe how you would access this feature?
Add a checkbox on the pipeline submission page to email the user when a pipeline is complete.  When the pipeline is complete the email controller should email the user that the pipeline is complete.

## Additional information
There should possibly be consideration to when there are many pipelines running for a user that they don't get spammed, but that could likely be done in a later enhancement to this feature.

Imported from GitLab issue #438.  Originally posted on 2017/03/27 03:08PM

### PR DESCRIPTION
**Pull request by deepsidhu85**
## Description of changes
What did you change in this pull request?  Provide a description of files changed, user interactions changed, etc.  Include how to test your changes.

Added ability for user to select if they would like to receive an email upon either pipeline completion or on error. 

Changes include adding a new column to the `analysis_submission table,` adding a checkbox to the pipeline launch UI, adding a email-pipeline-result mail template, call to the `EmailController` to send out an email if a pipeline completed successfully or is in error, adding an integration test, and adding this functionality to the user documentation

To test, run a development server with the production profile. Create a new user with your email address. Login with this user and create a new project, add a sample to this project and upload sequence files. Add this sample to the cart and select a pipeline. Select the "Receive Email upon Completion" checkbox. Launch the pipeline. Check to make sure you receive an email whether a pipeline succeeded or failed. 

## Related issue
Fixes #80 

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

* [x] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.
* [x] Tests added (or description of how to test) for any new features.
* [x] User documentation updated for UI or technical changes.

### CODE REVIEW

**Comment by tom114** [comment]
Need to add the new field to `ca/corefacility/bioinformatics/irida/sql/required-data.sql`

See error on startup:
```
21 Feb 2019 15:35:10,411 ERROR org.hibernate.tool.hbm2ddl.SchemaExport:457 - HHH000388: Unsuccessful: INSERT INTO `analysis_submission` (analysis_state, name, created_date, analysis_id, workflow_id, submitter, analysis_cleaned_state, priority, update_samples) VALUES ('COMPLETED', 'Analysis-1', '2014-08-06 10:01:02', 13, 'b7c8b437-3c41-485e-92e5-72b67e37959f', 2, 'NOT_CLEANED', 'MEDIUM', 0)....
```

File: src/test/java/ca/corefacility/bioinformatics/irida/service/impl/unit/AnalysisExecutionScheduledTaskImplTest.java [hunk 1]
```diff
@@ -243,6 +247,7 @@ public void testMonitorRunningAnalysesSuccessFinished() throws ExecutionManagerE
 
 		assertEquals(AnalysisState.FINISHED_RUNNING, analysisSubmission.getAnalysisState());
 		verify(analysisSubmissionRepository).save(analysisSubmission);
+		verify(emailController).sendPipelineStatusEmail(analysisSubmission);
```
Comment (tom114) [inline src/test/java/ca/corefacility/bioinformatics/irida/service/impl/unit/AnalysisExecutionScheduledTaskImplTest.java]: Can you add a test for "yes send email" and "don't send email"?

File: src/main/java/ca/corefacility/bioinformatics/irida/service/impl/AnalysisExecutionScheduledTaskImpl.java [hunk 2]
```diff
@@ -258,17 +266,21 @@ private void handleJobErrors(AnalysisSubmission analysisSubmission) {
 			AnalysisSubmission analysisSubmission) {
 		Future<AnalysisSubmission> returnedSubmission;
 
+		boolean final_workflow_status_set = false;
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/service/impl/AnalysisExecutionScheduledTaskImpl.java]: Fixed in commit 2c3c780199f0efb60c02fa3dd9f7dfc211d2d508

File: src/main/java/ca/corefacility/bioinformatics/irida/service/impl/AnalysisExecutionScheduledTaskImpl.java [hunk 3]
```diff
@@ -280,6 +292,11 @@ private void handleJobErrors(AnalysisSubmission analysisSubmission) {
 			analysisSubmission.setAnalysisState(AnalysisState.ERROR);
 			returnedSubmission = new AsyncResult<>(analysisSubmissionRepository.save(analysisSubmission));
 			handleJobErrors(analysisSubmission);
+			final_workflow_status_set = true;
+		}
+
+		if (final_workflow_status_set && analysisSubmission.getEmailPipelineResult()) {
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/service/impl/AnalysisExecutionScheduledTaskImpl.java]: Fixed in commit 2c3c780199f0efb60c02fa3dd9f7dfc211d2d508

File: src/main/java/ca/corefacility/bioinformatics/irida/service/impl/EmailControllerImpl.java [hunk 4]
```diff
@@ -237,13 +239,53 @@ public void sendNCBIUploadExceptionEmail(String adminEmailAddress, Exception roo
...
+				.getEmail());
+
+		Locale locale = LocaleContextHolder.getLocale();
+
+		final Context ctx = new Context(locale);
+		String pipelineStatus = submission.getAnalysisState()
+				.equals(AnalysisState.ERROR) ?
+				messageSource.getMessage("email.pipeline.ERROR", null, locale) :
+				messageSource.getMessage("email.pipeline.COMPLETED", null, locale);
+		String emailSubject =
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/service/impl/EmailControllerImpl.java]: Fixed in commit 2c3c780199f0efb60c02fa3dd9f7dfc211d2d508

File: src/test/java/ca/corefacility/bioinformatics/irida/service/impl/unit/AnalysisExecutionScheduledTaskImplTest.java [hunk 5]
```diff
@@ -243,18 +247,23 @@ public void testMonitorRunningAnalysesSuccessFinished() throws ExecutionManagerE
 
 		assertEquals(AnalysisState.FINISHED_RUNNING, analysisSubmission.getAnalysisState());
 		verify(analysisSubmissionRepository).save(analysisSubmission);
+		if (analysisSubmission.getEmailPipelineResult()) {
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/test/java/ca/corefacility/bioinformatics/irida/service/impl/unit/AnalysisExecutionScheduledTaskImplTest.java]: Fixed in commit 2c3c780199f0efb60c02fa3dd9f7dfc211d2d508

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/pipelines/PipelineController.java [hunk 6]
```diff
@@ -344,9 +347,11 @@ public String getSpecifiedPipelinePage(final Model model, Principal principal, L
 			model.addAttribute("addRefProjects", addRefList);
 			model.addAttribute("projects", projectList);
 			model.addAttribute("canUpdateSamples", canUpdateAllSamples);
-                        model.addAttribute("workflowName", workflowName);
+			model.addAttribute("workflowName", workflowName);
 			model.addAttribute("dynamicSourceRequired", description.requiresDynamicSource());
-			model.addAttribute("analysisType", flow.getWorkflowDescription().getAnalysisType());
+			model.addAttribute("analysisType", flow.getWorkflowDescription()
+					.getAnalysisType());
+			model.addAttribute("emailConfigured", emailController.isMailConfigured());
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/ria/web/pipelines/PipelineController.java]: This is all looking great.  Last change I can see is that if `isMailConfigured` is `false`, the error message in the logs is `E-mail host is not configured, unable to send password reset e-mails.`

Can you make a change in `WebEmailConfig` to update that message to remove the "password reset" part?  That likely should have been done in the past but we can do it here now.  Thanks.

Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/pipelines/PipelineController.java]: Fixed in 63c08683e9697673e7e99401101bfff14c658a05

**Pull request merged**